include/revng/Yield/Support/GraphLayout/Graphs.h

```diff
@@ -0,0 +1,140 @@
+                                  detail::InputEdge<EdgeDataType>,
+                                  false>;
+
+template<typename NodeDataType, typename EdgeDataType = Empty>
+class InputGraph
+  : public GenericGraph<InputNode<NodeDataType, EdgeDataType>, 16, true> {
+  using Base = GenericGraph<InputNode<NodeDataType, EdgeDataType>, 16, true>;
+
+public:
+  using Base::Base;
+};
+
+template<typename NodeDataType, typename EdgeDataType>
+using OutputNode = MutableEdgeNode<detail::OutputNode<NodeDataType>,
+                                   detail::OutputEdge<EdgeDataType>,
+                                   false>;
+
+template<typename NodeDataType, typename EdgeDataType>
+class OutputGraph
+  : public GenericGraph<OutputNode<NodeDataType, EdgeDataType>, 16, true> {
+  using Base = GenericGraph<OutputNode<NodeDataType, EdgeDataType>, 16, true>;
+
+public:
+  using Base::Base;
+};
+
+template<typename NodeDataType, typename EdgeDataType>
+struct LayoutableGraphTraits<InputGraph<NodeDataType, EdgeDataType> *> {
+  using GraphType = InputGraph<NodeDataType, EdgeDataType> *;
+  static_assert(SpecializationOfGenericGraph<std::remove_pointer_t<GraphType>>);
+  using LLVMTrait = llvm::GraphTraits<GraphType>;
+
+  static const Size &getNodeSize(typename LLVMTrait::NodeRef Node) {
+    return Node.Size;
+  }
+};
+
+template<typename NodeDataType, typename EdgeDataType>
+struct LayoutableGraphTraits<OutputGraph<NodeDataType, EdgeDataType> *> {
+  using GraphType = OutputGraph<NodeDataType, EdgeDataType> *;
+  static_assert(SpecializationOfGenericGraph<std::remove_pointer_t<GraphType>>);
+  using LLVMTrait = llvm::GraphTraits<GraphType>;
+
+  static const Size &getNodeSize(typename LLVMTrait::NodeRef Node) {
+    return Node->Size;
+  }
+
+  static void setNodePosition(typename LLVMTrait::NodeRef Node, Point &&Point) {
+    Node->Center = std::move(Point);
+  }
+  static void setEdgePath(typename LLVMTrait::EdgeRef Edge, Path &&Path) {
+    Edge.Label->Path = std::move(Path);
+  }
+};
+
+namespace detail {
+
+template<typename NodeDataType, typename EdgeDataType>
+OutputGraph<NodeDataType, EdgeDataType>
+convert(const InputGraph<NodeDataType, EdgeDataType> &Input) {
+  OutputGraph<NodeDataType, EdgeDataType> Result;
+
+  using InputNode = layout::InputNode<NodeDataType, EdgeDataType>;
+  using OutputNode = layout::OutputNode<NodeDataType, EdgeDataType>;
+
+  std::unordered_map<const InputNode *, OutputNode *> Lookup;
+  for (const InputNode *Node : Input.nodes()) {
+    OutputNode Output(*Node);
+    Output.Size = Node->Size;
+
+    auto [It, Success] = Lookup.emplace(Node,
+                                        Result.addNode(std::move(Output)));
+    revng_assert(Success);
+  }
+
+  for (const InputNode *From : Input.nodes())
+    for (auto [To, EdgeLabel] : From->successor_edges())
+      Lookup.at(From)->addSuccessor(Lookup.at(To),
+                                    EdgeDataType(std::move(*EdgeLabel)));
+
+  return Result;
+}
```

Comment:
Do we really need conversion? Given that the layout algorithm only works on its internal data structure, and we are already copying the input in there, can't we just update the input in-place at the end of the layout? This would save us an entire copy of the input.

Instead of input -> copy -> internal -> copy -> output we'd have inputoutput -> copy -> internal -> update -> inputoutput.

Reply:
Closing this as per mattermost discussion.